### ISSUE TITLE
Re-register service in consul if the service get's de-registered

### ISSUE THREAD

**Issue by mirwais27** [spring-cloud/spring-cloud-consul#197]
If a local consul agent is restarted (for what ever reason) then the local Spring Boot Apps (with spring-cloud-consul inside) does not a re-connect. Anything special that I missed to configure for having this?

**Comment by spencergibb** [comment]
I just tried it and works ok. There are no persistent connections to consul, so the next time it tries to connect, it should work. Are you seeing any errors?

**Comment by mirwais27** [comment]
We are starting the consul agent like that: `consul.exe agent -log-level info -data-dir .\consul-data -server -ui -ui-dir=.\consul-ui -bootstrap -advertise=127.0.0.1`. We configured also the retry feature in pur spring boot apps as mentioned here: http://cloud.spring.io/spring-cloud-static/spring-cloud.html#spring-cloud-consul-retry 
The issue is that from time to time we update or reinstall the consul agent - with deleting the **consul-data** folder. And if we start then the consul agent then the spring boot apps can't reconnect: The consul-ui `http://localhost:8500/` shows only the consul agent itself, but not the running services anymore. Workaround is to restart all spring boot apps as well. But if you have many of them then it's really annoying. 

It would be good if the spring boot apps re-register again automatically if consul is running again and they are not registered yet. Would make spring boot apps even more resilient.

**Comment by spencergibb** [comment]
> with deleting the consul-data folder
is the key statement there. So you're not asking for a reconnection, that works. You're asking for reregistration.

**Comment by mirwais27** [comment]
As you describe: re-register even if consul agent is started with a fresh empty consul-data folder would be really nice!

**Comment by spencergibb** [comment]
Knowing when to re-register is the problem then.

**Comment by pwielgolaski** [comment]
I have the same problem, when my consul instance is restarted, spring boot app is sending heartbeat bot service is missing and it get 500.
It would be nice to have re-register feature

**Comment by spencergibb** [comment]
how do you know a 500 should reregister?

**Comment by pwielgolaski** [comment]
Fair point, we can also check if error message contains information about missing check 
https://github.com/hashicorp/consul/blob/master/command/agent/acl.go#L372

**Comment by avarabyeu** [comment]
@spencergibb i believe, if health check attempt fails, we need to check whether service is still registered in consul. So we can re-register service if it got de-registered somehow.

**Comment by FlowerWrong** [comment]
The same issue.

**Comment by spencergibb** [comment]
Not sure I'm keen on adding something like this. This should be a function of the platform.

**Comment by dabarbanel** [comment]
Has there been any progress or a work around solution to this problem?

**Comment by spencergibb** [comment]
No, there hasn't. See my comments above.

**Comment by pwielgolaski** [comment]
"Simple things should be simple, complex things should be possible." ;)
@spencergibb could we make it possible ?
I see some reasoning behind not making it part of platform, but could we think how people that want to do it could do it?

**Comment by askaribragimov** [comment]
+1

**Comment by deevodavis71** [comment]
I just did some basic investigation into this issue, and the code below seems to work OK for me in a Spring Boot app. It just relies on the endpoint being manually poked by a Spring Actuator "/refresh" command after the Consul agent has been restarted (it even works if /refresh is called multiple times!)

`
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cloud.context.environment.EnvironmentChangeEvent;
import org.springframework.context.ApplicationListener;
import org.springframework.core.env.Environment;
import org.springframework.stereotype.Component;

import com.ecwid.consul.v1.ConsulClient;
import com.ecwid.consul.v1.agent.model.NewService;

@Component
public class RefreshConsul implements ApplicationListener<EnvironmentChangeEvent> {

	@Autowired
	Environment environment;

	@Override
	public void onApplicationEvent(EnvironmentChangeEvent event) {

		System.out.println("\n\nREFRESHED!!!!\n\n");

		try {

			// do stuff
			ConsulClient client = new ConsulClient("localhost");

			// Get the port
			String port = environment.getProperty("local.server.port");
			if (port == null)
				return;

			// Get the app name
			String appName = environment.getProperty("spring.application.name");

			// Fix underscores
			appName = appName.replace('_', '-');

			// register new service
			NewService newService = new NewService();
			newService.setId(appName + "-" + port);
			newService.setName(appName);
			newService.setPort(Integer.parseInt(port));

			// Add the health check
			NewService.Check serviceCheck = new NewService.Check();
			serviceCheck.setHttp("http://localhost:" + port + "/health");
			serviceCheck.setInterval("10s");

			// register the health check
			newService.setCheck(serviceCheck);

			// register the service
			client.agentServiceRegister(newService);

		} catch (NumberFormatException e) {

			// Oops
			e.printStackTrace();

		}
	}

}

`

**Comment by gitrc** [comment]
+1 this would be extremely helpful as we end up restarting our spring applications when redeploying/upgrading consul agents.

**Comment by perdonare** [comment]
+1

**Comment by SystemOutPrint** [comment]
+1

**Comment by deripas** [comment]
+1

**Comment by laurikimmel** [comment]
Encountered same problem - restarting consul will clean the service registry and render service discovery broken. Services have to be restarted in order to register them again.

**Comment by rotilho** [comment]
I addressed this issue adding a health check for consul registration so my orchestrator is able to restarted it after the application is unhealth for while.

I first attempted to create a reconnect mechanism but spring consul have protected the registration method.

**Comment by mohkharma** [comment]
Any update on  how to solve this case?

**Comment by michelfigueiredojava** [comment]
> I addressed this issue adding a health check for consul registration so my orchestrator is able to restarted it after the application is unhealth for while.
> 
> I first attempted to create a reconnect mechanism but spring consul have protected the registration method.

Hello @rotilho I am looking for a workaround for this issue, could you please explain how did you do that ?

**Comment by rotilho** [comment]
I just added a health check and let my orchestrator handle the situation when the application is unhealth

`@Slf4j
@Component
@RequiredArgsConstructor
@ConditionalOnProperty(value = "spring.cloud.consul.enabled")
public class ConsulRegistrationHealthIndicator extends AbstractHealthIndicator {
    public static final String APPLICATION_NAME_KEY = "spring.application.name";
    public static final String HOSTNAME_KEY = "spring.cloud.consul.discovery.hostname";
    public static final String PORT_KEY = "spring.cloud.consul.discovery.port";

    private final Environment environment;
    private final ConsulClient client;


    protected void doHealthCheck(Health.Builder builder) {
        String applicationName = environment.getRequiredProperty(APPLICATION_NAME_KEY);
        String address = environment.getRequiredProperty(HOSTNAME_KEY);
        Integer servicePort = environment.getRequiredProperty(PORT_KEY, Integer.class);

        try {
            Response<List<CatalogService>> services = this.client.getCatalogService(applicationName, QueryParams.DEFAULT);
            Optional<CatalogService> catalogService = services.getValue().stream()
                    .filter(service -> service.getServiceAddress().equals(address))
                    .filter(service -> service.getServicePort().equals(servicePort))
                    .findAny();

            if (catalogService.isPresent()) {
                builder.up();
            } else {
                builder.down();
            }
        } catch (Exception e) {
            log.error("Couldn't query catalog service for {}", applicationName, e);
            builder.down(e);
        }

    }
}`

**Comment by michelfigueiredojava** [comment]
> I just added a health check and let Marathon restart my unhealth application. It's not optimal since my application may stay offline for few seconds but in my case is acceptable.
> 
> ```
> @Slf4j
> @Component
> @RequiredArgsConstructor
> @ConditionalOnProperty(value = "spring.cloud.consul.enabled")
> public class ConsulRegistrationHealthIndicator extends AbstractHealthIndicator {
>     public static final String APPLICATION_NAME_KEY = "spring.application.name";
>     public static final String HOSTNAME_KEY = "spring.cloud.consul.discovery.hostname";
>     public static final String PORT_KEY = "spring.cloud.consul.discovery.port";
> 
>     private final Environment environment;
>     private final ConsulClient client;
> 
> 
>     protected void doHealthCheck(Health.Builder builder) {
>         String applicationName = environment.getRequiredProperty(APPLICATION_NAME_KEY);
>         String address = environment.getRequiredProperty(HOSTNAME_KEY);
>         Integer servicePort = environment.getRequiredProperty(PORT_KEY, Integer.class);
> 
>         try {
>             Response<List<CatalogService>> services = this.client.getCatalogService(applicationName, QueryParams.DEFAULT);
>             Optional<CatalogService> catalogService = services.getValue().stream()
>                     .filter(service -> service.getServiceAddress().equals(address))
>                     .filter(service -> service.getServicePort().equals(servicePort))
>                     .findAny();
> 
>             if (catalogService.isPresent()) {
>                 builder.up();
>             } else {
>                 builder.down();
>             }
>         } catch (Exception e) {
>             log.error("Couldn't query catalog service for {}", applicationName, e);
>             builder.down(e);
>         }
> 
>     }
> }
> ```

thanks, but in my case I can't use this solution. 

Anyone else has another workaround please ?

**Comment by havelh** [comment]
> > I just added a health check and let Marathon restart my unhealth application. It's not optimal since my application may stay offline for few seconds but in my case is acceptable.
> > ```
> > @Slf4j
> > @Component
> > @RequiredArgsConstructor
> > @ConditionalOnProperty(value = "spring.cloud.consul.enabled")
> > public class ConsulRegistrationHealthIndicator extends AbstractHealthIndicator {
> >     public static final String APPLICATION_NAME_KEY = "spring.application.name";
> >     public static final String HOSTNAME_KEY = "spring.cloud.consul.discovery.hostname";
> >     public static final String PORT_KEY = "spring.cloud.consul.discovery.port";
> > 
> >     private final Environment environment;
> >     private final ConsulClient client;
> > 
> > 
> >     protected void doHealthCheck(Health.Builder builder) {
> >         String applicationName = environment.getRequiredProperty(APPLICATION_NAME_KEY);
> >         String address = environment.getRequiredProperty(HOSTNAME_KEY);
> >         Integer servicePort = environment.getRequiredProperty(PORT_KEY, Integer.class);
> > 
> >         try {
> >             Response<List<CatalogService>> services = this.client.getCatalogService(applicationName, QueryParams.DEFAULT);
> >             Optional<CatalogService> catalogService = services.getValue().stream()
> >                     .filter(service -> service.getServiceAddress().equals(address))
> >                     .filter(service -> service.getServicePort().equals(servicePort))
> >                     .findAny();
> > 
> >             if (catalogService.isPresent()) {
> >                 builder.up();
> >             } else {
> >                 builder.down();
> >             }
> >         } catch (Exception e) {
> >             log.error("Couldn't query catalog service for {}", applicationName, e);
> >             builder.down(e);
> >         }
> > 
> >     }
> > }
> > ```
> 
> thanks, but in my case I can't use this solution.
> 
> Anyone else has another workaround please ?

I have extended the ConsulRegistrationHealthIndicator  from above and all my services do a reregister after the consul agent is restarted:

```
/*
 * this class checks periodically, if the application is in the catalog,
 * and if not he tries to reregister it self.
 */

@Configuration
@ConditionalOnProperty(value = "spring.cloud.consul.enabled", matchIfMissing = true)
@ConditionalOnConsulEnabled
@ConditionalOnDiscoveryEnabled
public class ConsulRegistrationHealthIndicator implements SmartLifecycle {
    Logger logger = LoggerFactory.getLogger(ConsulRegistrationHealthIndicator.class);

	
	public static final String APPLICATION_NAME_KEY = "spring.application.name";
    public static final String APPLICATION_PORT_KEY = "server.port";

    private final Environment environment;
    private final ConsulClient client;

	@Autowired
	AbstractAutoServiceRegistration serviceRegistration;         	

	private final TaskScheduler taskScheduler = getTaskScheduler();

	private final AtomicBoolean running = new AtomicBoolean(false);

	private ScheduledFuture<?> watchFuture;	

	
	public ConsulRegistrationHealthIndicator(Environment environment, ConsulClient client) {
		super();

		this.environment = environment;
		this.client = client;
	}


	@Bean
	@ConditionalOnMissingBean
	public ConsulRegistrationHealthIndicator consulRegistrationHealthIndicator(
			Environment discoveryProperties, ConsulClient consulClient) {
		
		return new ConsulRegistrationHealthIndicator(discoveryProperties, consulClient);
	}

	private static ThreadPoolTaskScheduler getTaskScheduler() {
		ThreadPoolTaskScheduler taskScheduler = new ThreadPoolTaskScheduler();
		taskScheduler.initialize();
		return taskScheduler;
	}

	@Override
	public boolean isAutoStartup() {
		return true;
	}

	@Override
	public void stop(Runnable callback) {
		this.stop();
		callback.run();
	}

	@Override
	public void start() {
		if (this.running.compareAndSet(false, true)) {
			
			this.watchFuture = this.taskScheduler.scheduleWithFixedDelay(
					this::doCheck, 13000);
		}
	}

	@Override
	public void stop() {
		if (this.running.compareAndSet(true, false) && this.watchFuture != null) {
			this.watchFuture.cancel(true);
		}
	}

	@Override
	public boolean isRunning() {
		return false;
	}

	@Override
	public int getPhase() {
		return 0;
	}

	public void doCheck() {
        String applicationName = environment.getRequiredProperty(APPLICATION_NAME_KEY);
        Integer myServicePort = environment.getRequiredProperty(APPLICATION_PORT_KEY, Integer.class);
        
        InetAddress myIp;
        try {
            myIp = InetAddress.getLocalHost();
        } catch (UnknownHostException e) {
        	return;
        }
        
        String myHostname = myIp.getCanonicalHostName();
        
    	logger.info("my current IP address : {}  hostname : {}", myIp, myHostname);
        
    	logger.info("health check my self {}", applicationName);

    	if(!serviceRegistration.isRunning())  {
        	logger.info("i am {} not yet running", applicationName);
        	// nothing to do
        	return;
    	}
    	
        try {
            Response<List<CatalogService>> services = this.client.getCatalogService(applicationName, QueryParams.DEFAULT);
            
            Optional<CatalogService> catalogService = services.getValue().stream()
                    .filter(service -> service.getServiceAddress().equals(myHostname) && service.getServicePort().equals(myServicePort))
                    .findAny();

            if (!catalogService.isPresent()) {
            	// reregister my self
            	logger.warn("i am {} not in the registry, do reregister", applicationName);
                
            	serviceRegistration.stop();
            	serviceRegistration.start();
            } else {
            	logger.info("i am {} in the registry, nothing to do", applicationName);
            }
        } catch (Exception e) {
        	logger.error("Couldn't query catalog service for {}", applicationName, e);
        }

    }
}
```

**Comment by hetianyi** [comment]
@pwielgolaski I encountered a similar problem. When both the consul server and the application are ok, but the network has a problem. At this time, the consul server will remove my application registration, when the network is restored again, my application will not be able to automatically registered, and keeps throwing Exception : ```OperationException(statusCode=500, statusMessage='Internal Server Error', statusContent='CheckID "service:xxx" does not have associated TTL')```, This is unexpected!

This is the solution I tried:

[https://github.com/spring-cloud/spring-cloud-consul/compare/master...hetianyi:master](https://github.com/spring-cloud/spring-cloud-consul/compare/master...hetianyi:master)

```java
try {
	TtlScheduler.this.client.agentCheckPass(this.checkId);
}
catch (Exception e) {
	// try to re-register service, this may fail again.
	try {
		TtlScheduler.this.client.agentServiceRegister(reg.getService(),
				TtlScheduler.this.properties.getAclToken());
		if (log.isDebugEnabled()) {
			log.debug("Agent check failed for " + this.checkId + ", re-registered");
		}
	} finally {
		throw e;
	}
}
```
> agentServiceRegister() seems to be an idempotent operation. I'm not very sure. Check this [https://github.com/hashicorp/consul/pull/4905](https://github.com/hashicorp/consul/pull/4905)

Rebuild "spring-cloud-consul-discovery" module and replace it:
```xml
<dependency>
    <groupId>org.springframework.cloud</groupId>
    <artifactId>spring-cloud-starter-consul-discovery</artifactId>
    <exclusions>
        <exclusion>
            <groupId>org.springframework.cloud</groupId>
            <artifactId>spring-cloud-consul-discovery</artifactId>
        </exclusion>
    </exclusions>
</dependency>
<dependency>
    <groupId>org.springframework.cloud</groupId>
    <artifactId>spring-cloud-consul-discovery</artifactId>
    <version>3.0.0.BUILD-SNAPSHOT</version>
</dependency>
```
This works for me!

**Comment by ntchern** [comment]
+1

I have a healthy Consul instance (even though it was restarted/recreated, why should I care?), I have a healthy microservice instance that tries to re-establish itself into that Consul instance, but fails to do so. It doesn't feel right. Isn't it the whole point of service discovery?

"Service discovery is the process of automatically detecting devices and services on a network." 

I am not convinced that I have to maintain consul-data folder in order to make this process to work seamlessly. This extra step also doesn't feel right in the era of docker and kubernetes, when everything is disposable as much as possible.

Can it please be addressed so it'll work without having to implement workaround or switching to another product?

**Comment by spencergibb** [comment]
With the replace existing checks parameter, I feel like this is safe enough Todo. It should be opt in to a scheduled method that simply reregisters.

PRs welcome

**Comment by making** [comment]
I will make a PR to support an option of re-registration.

**Comment by ghimisradu** [comment]
Any update on this, when will it be available on maven central?

**Comment by hanct** [comment]
Yi, has this been resolved? i am currently using Spring Boot: v2.5.1 and using org.springframework.cloud:spring-cloud-starter-consul-discovery.  But still the re-register service is not working.

i have raised the issue here https://github.com/spring-cloud/spring-cloud-consul/issues/727,  but no response
